library/alloc/src/collections/btree/map.rs

```diff
@@ -223,30 +224,37 @@ where
     type Key = K;
 
     fn get(&self, key: &Q) -> Option<&K> {
-        match search::search_tree(self.root.as_ref()?.as_ref(), key) {
+        let root_node = self.root.as_ref()?.node_as_ref();
+        match search::search_tree(root_node, key) {
             Found(handle) => Some(handle.into_kv().0),
             GoDown(_) => None,
         }
     }
 
     fn take(&mut self, key: &Q) -> Option<K> {
-        match search::search_tree(self.root.as_mut()?.as_mut(), key) {
-            Found(handle) => Some(
-                OccupiedEntry { handle, length: &mut self.length, _marker: PhantomData }
-                    .remove_kv()
-                    .0,
-            ),
+        let (maplease, search_result) = DormantMutRef::maybe_new(self, |this| {
+            let root = unsafe { (*this.as_ptr()).root.as_mut() }?;
```

Comment:
With code like this, I wonder... can't we make `this` a reference instead of a `NonNull`?

Reply:
But [they told me](https://github.com/rust-lang/rust/pull/74437) (or I understood) that dereferencing will invalidate the pointer? I suppose you say, get the pointer after the closure got the part... but then it's a double borrow?

Reply:
What I mean is, `DormantMutRef::new` would turn the input ref into a `NonNull`, store than `NonNull`, and then turn the `NonNull` back into a ref that is passed to the closure. That way, the ptr in the closure is *derived from* the raw ptr that lies dormant.

That's basically what you are already doing, isn't it?

Reply:
Well, that's actually what I was trying to _avoid_ :confused: because I thought turning the `NonNull` into a ref and dereferencing would invalidate the pointer itself, and all pointers to parts, and we need to preserve both: the pointer itself to eventually get back to the map, and the "reference to a part" is actually a handle disguising another `NonNull`.

Reply:
Ok I'm stupid, there are no "pointers to parts" around yet, we're only getting ready to calculate them. So dereferencing the map reference does not invalidate the pointer to the same thing? Probably that's already explicitly written somewhere.

Reply:
Using a pointer invalidates all *child* and *sibling* pointers. But here `self` is a *parent* pointer of the `NonNull`.

Reply:
I don't remember seeing the sibling wording and can't quite figure out what it means.

> `self` is a parent pointer of the `NonNull`.

So that means the `NonNull` was derived from the `&mut self`, and not the other way around?

I thought that the code I put in yesterday:
```
let sub = compute_sub(unsafe { &mut *ptr.as_ptr() });`
```
and not simply:
```
let sub = compute_sub(t);`
```
is what you meant or in other words necessary to respect CAD97's wording "&mut S is a unique reference to the entirety of S. When you use it, that invalidates any raw references derived from it". But if use the simple code, and add in some intermediate functions that otherwise seem to motivate Miri to check more, Miri still doesn't report anything.

Reply:
Well, Miri is clearly suffering from the virus or the heath... it no longer complains about [this code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=aafa121751f188b9f77639f1825fc387) that it complained about a month ago. Or was it actually wrong before? Now I really don't know anymore if or when dereferencing a `&mut` invalidates a pointer to the same area. Miri certainly still invalidates a pointer to a part of it, i.e. the ignored tests fixed by #73971 still fail.

Reply:
> I don't remember seeing the sibling wording and couldn't yet figure out what it means.

Fair, I didn't define it. I am trying different terminology to see what works best. ;)
These are siblings:
```rust
fn foo(parent: &mut i32) {
  let sibling1 = &mut *parent;
  let sibling2 = &mut *parent;
}
```
The both reborrow from the same parent reference.

> it no longer complains about this code that it complained about a month ago. 

I don't think there were any changes in Miri that would affect this code...

>  Now I really don't know anymore if or when dereferencing a &mut invalidates a pointer to the same area.

Dereferencing a `ptr` of type `&mut` invalidates all pointers to that area *except the pointers from which `ptr` was derived*, i.e., except for `ptr`'s "ancestors".

Reply:
I'm suffering from the heath. Miri does still complain, and is still silent when the function is inlined.

Reply:
[This is the API I had in mind](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0fe02132e85db22089141fd0fde3c877). I also adjusted the doc comments.

Reply:
Ah you are right, my `new` contains a safety hole. It can be fixed with a lifetime bound, but I am not sure if the API is still useful with that bound:
```rust
    pub fn new<'n>(t: &'t mut T) -> (Self, &'n mut T)
        where 't: 'n
```

Reply:
Right, I understood, for 22 minutes, that we _can_ expose and use the entire reference. Now I don't see what it is that I'm right about. If you have a real reference, the borrow checker makes sure you don't borrow from it more than once?

Rest assured I do try to read the documentation, but understanding it is easier if you can ask Miri to confirm when the UB occurs exactly. But even if Miri were perfect, there's still plenty to confuse me: over the last hours I've been editing the wrong file, running the wrong command, with the wrong toolchain (the latter doesn't matter, but if you're convinced Miri changed over time, it helps to test it right).

So now I'm pretty confident that using the line:

```
let sub = compute_sub(t);
```

is indeed wrong and I'll try to get Miri to make a stand against it.

Reply:
[I extended your code with my test code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1aa15224295bfddf85bc64a8d9d981c0). I understand the safety hole in the original, but not why you consider that limits its usefulness: to my knowledge, `'t` is an era spanning the whole story here, the lifetime of the borrow of `BTreeMap::drain_filter` or `BTreeMap::entry`, how can that be limiting? I'm not entirely sure why you don't just let `new` return a `&'t mut T`, but I can imagine situations where a short-lived `'n` gets things compiled.

Reply:
> Rest assured I do try to read the documentation, but understanding it is easier if you can ask Miri to confirm when the UB occurs exactly. 

Absolutely.
I just hope at some point I can get around to make Miri track raw pointers better, I did not quite anticipate how much confusion this causes...

> to my knowledge, 't is an era spanning the whole story here, the lifetime of the borrow of BTreeMap::drain_filter or BTreeMap::entry, how can that be limiting?

I had no idea what `'t` is. If it really is a long lifetime like that, even better. :)

>  I'm not entirely sure why you don't just let new return a &'t mut T, but I can imagine situations where a short-lived 'n gets things compiled. Plus it's darn useful just to document.

It's actually equivalent to returning `'t` (since `'t` is a valid choice for `'n`). The reason I left `'n` is that it gives me a name to refer to in the docs. I don't know if that's a good thing to do or not...

Reply:
I thought earlier today that when lifetimes are discussed, we always implicitly mean the compile-time lifetime. And this Dormant thing attaches a shorter run-time lifetime concept to the lifetime identifier `n (or `b now).

> since `'t` is a valid choice for `'n`

Apparently it is. Even in this code, where I first thought that the `drop` would shorten the lifetime `n`:

```
    let (ptr_s, ref_s) = DormantMutRef::new(&mut s);
    ref_s.a += 1;
    drop(ref_s);
    unsafe { ptr_s.awaken() }.b += 1;
```

But the compiler doesn't complain if I take away its toy `'n` and force it to be `'t`. Now I think I realize that the drop does not influence that lifetime, it just ends the lifetime of the `ref_s` variable. So does that mean it's good or bad to speak of a shorter run-time lifetime? To be continued...

Reply:
`drop` is just a normal function, why would it have any special effect on lifetimes?

"run-time lifetime" is not really a thing. Lifetimes are compile-time fiction. The closest runtime analogue we have is Stacked Borrows and the question "for how long does this pointer's tag remain on the stack".

Reply:
> "run-time lifetime" is not really a thing.

Well you made it a thing, by commenting "lifetine `'n` from `new` must be over" (sic). If the actual compile-time lifetime `'n` is not over (cannot be over), what kind of lifetime is this about? I guess it's the use of the reference, the effective lifetime, run-time lifetime...?

I'm beginning to feel using the word lifetime here doesn't help (me) and it's better to make everything `'t` and explain that that is just a compile-time thing we have to be vague about.

Reply:
> Well you made it a thing, by commenting "lifetine 'n from new must be over" (sic). If the actual compile-time lifetime 'n is not over (cannot be over), what kind of lifetime is this about? I guess it's the use of the reference, the effective lifetime, run-time lifetime...?

Fair, my terminology hasn't been as precise as it should be.

What I meant is that the reference returned by `new`, and all its descendants (i.e., all other pointers and references derived from it), may not be used any more after `awaken` is called.